api/src/main/java/io/opentelemetry/trace/propagation/B3Propagator.java

```diff
@@ -0,0 +1,204 @@
+
+    if (singleHeader) {
+      return getSpanContextFromSingleHeader(carrier, getter);
+    } else {
+      return getSpanContextFromMultipleHeaders(carrier, getter);
+    }
+  }
+
+  @SuppressWarnings("StringSplitter")
+  private static <C /*>>> extends @NonNull Object*/> SpanContext getSpanContextFromSingleHeader(
+      C carrier, Getter<C> getter) {
+    String value = getter.get(carrier, COMBINED_HEADER);
+    if (StringUtils.isNullOrEmpty(value)) {
+      logger.info(
+          "Missing or empty combined header: "
+              + COMBINED_HEADER
+              + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    // must have between 2 and 4 hyphen delimieted parts:
+    //   traceId-spanId-sampled-parentSpanId (last two are optional)
+    // NOTE: we do not use parentSpanId
+    String[] parts = value.split(COMBINED_HEADER_DELIMITER);
+    if (parts.length < 2 || parts.length > 4) {
+      logger.info(
+          "Invalid combined header '" + COMBINED_HEADER + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String traceId = parts[0];
+    if (!isTraceIdValid(traceId)) {
+      logger.info(
+          "Invalid TraceId in B3 header: " + COMBINED_HEADER + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String spanId = parts[1];
+    if (!isSpanIdValid(spanId)) {
+      logger.info(
+          "Invalid SpanId in B3 header: " + COMBINED_HEADER + ". Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String sampled = parts.length == 3 ? parts[2] : null;
+
+    return buildSpanContext(traceId, spanId, sampled);
+  }
+
+  private static <C /*>>> extends @NonNull Object*/> SpanContext getSpanContextFromMultipleHeaders(
+      C carrier, Getter<C> getter) {
+    String traceId = getter.get(carrier, TRACE_ID_HEADER);
+    if (!isTraceIdValid(traceId)) {
+      logger.info(
+          "Invalid TraceId in B3 header: "
+              + TRACE_ID_HEADER
+              + "'. Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String spanId = getter.get(carrier, SPAN_ID_HEADER);
+    if (!isSpanIdValid(spanId)) {
+      logger.info(
+          "Invalid SpanId in B3 header: " + SPAN_ID_HEADER + "'. Returning INVALID span context.");
+      return SpanContext.getInvalid();
+    }
+
+    String sampled = getter.get(carrier, SAMPLED_HEADER);
+    return buildSpanContext(traceId, spanId, sampled);
+  }
+
+  private static SpanContext buildSpanContext(String traceId, String spanId, String sampled) {
+    try {
+      TraceFlags traceFlags =
+          TraceFlags.builder()
+              .setIsSampled(
+                  TRUE_INT.equals(sampled)
+                      || Boolean.parseBoolean(sampled)) // accept either "1" or "true"
+              .build();
+
+      return SpanContext.createFromRemoteParent(
+          TraceId.fromLowerBase16(traceId, 0),
```

Comment:
This requires traceId (string) to be 32 hex characters and we only check that the size is not greater than MAX_TRACE_ID_LENGTH.

Reply:
Pushed change to left pad TraceIds upto max with leading 0's. I added padLeft to `StringUtils`. Let me know what you think.

Reply:
As this was fixed along with a test (see https://github.com/open-telemetry/opentelemetry-java/pull/979/files#diff-004fb9fadd7caedae3844dfe8b040186R137) I'm marking this as resolved.